### PR TITLE
[SPARK-32280][SPARK-32372][SQL]  ResolveReferences.dedupRight should only rewrite attributes for ancestor nodes of the conflict plan

### PR DESCRIPTION
**Pull request by Ngone51**
### What changes were proposed in this pull request?

This PR refactors `ResolveReferences.dedupRight` to make sure it only rewrite attributes for ancestor nodes of the conflict plan.

### Why are the changes needed?

This is a bug fix.

```scala
sql("SELECT name, avg(age) as avg_age FROM person GROUP BY name")
  .createOrReplaceTempView("person_a")
sql("SELECT p1.name, p2.avg_age FROM person p1 JOIN person_a p2 ON p1.name = p2.name") 
  .createOrReplaceTempView("person_b")
sql("SELECT * FROM person_a UNION SELECT * FROM person_b")
  .createOrReplaceTempView("person_c")
sql("SELECT p1.name, p2.avg_age FROM person_c p1 JOIN person_c p2 ON p1.name = p2.name").show()
```
When executing the above query, we'll hit the error:

```scala
[info]   Failed to analyze query: org.apache.spark.sql.AnalysisException: Resolved attribute(s) avg_age#231 missing from name#223,avg_age#218,id#232,age#234,name#233 in operator !Project [name#233, avg_age#231]. Attribute(s) with the same name appear in the operation: avg_age. Please check if the right attribute(s) are used.;;
...
```

The plan below is the problematic plan which is the right plan of a `Join` operator. And, it has conflict plans comparing to the left plan. In this problematic plan, the first `Aggregate` operator (the one under the first child of `Union`) becomes a conflict plan compares to the left one and has a rewrite attribute pair as  `avg_age#218` -> `avg_age#231`. With the current `dedupRight` logic, we'll first replace this `Aggregate` with a new one, and then rewrites the attribute `avg_age#218` from bottom to up. As you can see, projects with the attribute `avg_age#218` of the second child of the `Union` can also be replaced with `avg_age#231`(That means we also rewrite attributes for non-ancestor plans for the conflict plan). Ideally, the attribute `avg_age#218` in the second `Aggregate` operator (the one under the second child of `Union`) should also be replaced. But it didn't because it's an `Alias` while we only rewrite `Attribute` yet. Therefore, the project above the second `Aggregate` becomes unresolved.

```scala
:    
:
+- SubqueryAlias p2
   +- SubqueryAlias person_c
      +- Distinct
         +- Union
            :- Project [name#233, avg_age#231]
            :  +- SubqueryAlias person_a
            :     +- Aggregate [name#233], [name#233, avg(cast(age#234 as bigint)) AS avg_age#231]
            :        +- SubqueryAlias person
            :           +- SerializeFromObject [knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$Person, true])).id AS id#232, staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$Person, true])).name, true, false) AS name#233, knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$Person, true])).age AS age#234]
            :              +- ExternalRDD [obj#165]
            +- Project [name#233 AS name#227, avg_age#231 AS avg_age#228]
               +- Project [name#233, avg_age#231]
                  +- SubqueryAlias person_b
                     +- !Project [name#233, avg_age#231]
                        +- Join Inner, (name#233 = name#223)
                           :- SubqueryAlias p1
                           :  +- SubqueryAlias person
                           :     +- SerializeFromObject [knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$Person, true])).id AS id#232, staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$Person, true])).name, true, false) AS name#233, knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$Person, true])).age AS age#234]
                           :        +- ExternalRDD [obj#165]
                           +- SubqueryAlias p2
                              +- SubqueryAlias person_a
                                 +- Aggregate [name#223], [name#223, avg(cast(age#224 as bigint)) AS avg_age#218]
                                    +- SubqueryAlias person
                                       +- SerializeFromObject [knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$Person, true])).id AS id#222, staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$Person, true])).name, true, false) AS name#223, knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$Person, true])).age AS age#224]
                                          +- ExternalRDD [obj#165]
```

### Does this PR introduce _any_ user-facing change?

Yes, users would no longer hit the error after this fix.

### How was this patch tested?

Added test.

### CODE REVIEW

**Comment by Ngone51** [comment]
ping @cloud-fan @maropu @imback82 Please take a look, thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -1237,20 +1249,44 @@ class Analyzer(
...
-        // the reason that `conflictPlans` are also collected in pre-order.
-        right transformDown {
-          case r => conflictPlanMap.getOrElse(r, r)
-        } transformUp {
-          case other => other transformExpressions {
+        rewritePlan(right, conflictPlans.toMap)._1
+      }
+    }
+
+    private def rewritePlan(plan: LogicalPlan, conflictPlanMap: Map[LogicalPlan, LogicalPlan])
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1256]: This rewriting is much more surgical than before, +1

**Comment by cloud-fan** [comment]
cc @viirya

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -1237,20 +1249,44 @@ class Analyzer(
...
+      val attrMapping = new mutable.ArrayBuffer[(Attribute, Attribute)]()
+      if (conflictPlanMap.contains(plan)) {
+        // If the plan is the one that conflict the with left one, we'd
+        // just replace it with the new plan and collect the rewrite
+        // attributes for the parent node.
+        val newRelation = conflictPlanMap(plan)
+        attrMapping ++= plan.output.zip(newRelation.output)
+        newRelation -> attrMapping
+      } else {
+        var newPlan = plan.mapChildren { child =>
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1267]: nit: `val` is better.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -1237,20 +1249,44 @@ class Analyzer(
...
-        right transformDown {
-          case r => conflictPlanMap.getOrElse(r, r)
-        } transformUp {
-          case other => other transformExpressions {
+        rewritePlan(right, conflictPlans.toMap)._1
+      }
+    }
+
+    private def rewritePlan(plan: LogicalPlan, conflictPlanMap: Map[LogicalPlan, LogicalPlan])
+      : (LogicalPlan, mutable.ArrayBuffer[(Attribute, Attribute)]) = {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1257]: nit: In the most cases of recursive calls, a return type would be `Seq` instead of `ArrayBuffer`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 4]
```diff
@@ -1237,20 +1249,44 @@ class Analyzer(
...
+
+        if (attrMapping.isEmpty) {
+          newPlan -> attrMapping
+        } else {
+          assert(!attrMapping.groupBy(_._1.exprId)
+            .exists(_._2.map(_._2.exprId).distinct.length > 1),
+            "Found duplicate rewrite attributes")
+          val attributeRewrites = AttributeMap(attrMapping)
+          // rewrite the attributes of parent node
+          newPlan = newPlan.transformExpressions {
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1283]: Oh, I see. This looks more clear. +1 for this change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 5]
```diff
@@ -1192,11 +1192,23 @@ class Analyzer(
...
+        case _ @ Project(projectList, _)
+          if findAliases(projectList).size == projectList.size =>
+          Nil
+
         case oldVersion @ Aggregate(_, aggregateExpressions, _)
             if findAliases(aggregateExpressions).intersect(conflictingAttributes).nonEmpty =>
           Seq((oldVersion, oldVersion.copy(
             aggregateExpressions = newAliases(aggregateExpressions))))
 
+        case _ @ Aggregate(_, aggregateExpressions, _)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1208]: Same reason as above? Add a simple comment too?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 6]
```diff
@@ -1192,11 +1192,23 @@ class Analyzer(
             if findAliases(projectList).intersect(conflictingAttributes).nonEmpty =>
           Seq((oldVersion, oldVersion.copy(projectList = newAliases(projectList))))
 
+        // We don't need to search child plan recursively if the projectList of a Project
+        // is only composed of Alias and doesn't contain any conflicting attributes.
+        // Because, even if the child plan has some conflicting attributes, the attributes
+        // will be aliased to non-conflicting attributes by the Project at the end.
+        case _ @ Project(projectList, _)
+          if findAliases(projectList).size == projectList.size =>
+          Nil
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1201]: Don't we need to put this before previous `Project` pattern?

**Review by viirya — COMMENTED** [review]
Looks good.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 3 above)
Comment (Ngone51) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1257]: We can return `Seq` here. But I think it's not needed if the  `attrMapping` below is still an `Arraybuffer`. Or, you mean the `attrMapping` could also be `Seq`? (Please also note the `attrMapping` has to be an mutable collection anyway.)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 5 above)
Comment (Ngone51) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1208]: Yes, sure.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 6 above)
Comment (Ngone51) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1201]: I think we don't. It's mutually exclusive with the previous one as it has an implicit requirement ( `findAliases(projectList).intersect(conflictingAttributes).isEmpty`). Therefore, the actual requirement here is:
`findAliases(projectList).intersect(conflictingAttributes).isEmpty && findAliases(projectList).size == projectList.size`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 3 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1258]: Yea, cosmetic changes though, I thought it like this;
```
@@ -1254,16 +1254,15 @@ class Analyzer(
     }
 
     private def rewritePlan(plan: LogicalPlan, conflictPlanMap: Map[LogicalPlan, LogicalPlan])
-      : (LogicalPlan, mutable.ArrayBuffer[(Attribute, Attribute)]) = {
-      val attrMapping = new mutable.ArrayBuffer[(Attribute, Attribute)]()
+      : (LogicalPlan, Seq[(Attribute, Attribute)]) = {
       if (conflictPlanMap.contains(plan)) {
         // If the plan is the one that conflict the with left one, we'd
         // just replace it with the new plan and collect the rewrite
         // attributes for the parent node.
         val newRelation = conflictPlanMap(plan)
-        attrMapping ++= plan.output.zip(newRelation.output)
-        newRelation -> attrMapping
+        newRelation -> plan.output.zip(newRelation.output)
       } else {
+        val attrMapping = new mutable.ArrayBuffer[(Attribute, Attribute)]()
         var newPlan = plan.mapChildren { child =>
           // If not, we'd rewrite child plan recursively until we find the
           // conflict node or reach the leaf node.
```

Comment (Ngone51) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1258]: I see. I can follow your suggestion. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 7]
```diff
@@ -1237,20 +1250,42 @@ class Analyzer(
...
+    }
+
+    private def rewritePlan(plan: LogicalPlan, conflictPlanMap: Map[LogicalPlan, LogicalPlan])
+      : (LogicalPlan, Seq[(Attribute, Attribute)]) = {
+      if (conflictPlanMap.contains(plan)) {
+        // If the plan is the one that conflict the with left one, we'd
+        // just replace it with the new plan and collect the rewrite
+        // attributes for the parent node.
+        val newRelation = conflictPlanMap(plan)
+        newRelation -> plan.output.zip(newRelation.output)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1264]: nit: `conflictPlanMap(plan) -> plan.output.zip(newRelation.output)`

**Comment by ScrapCodes** [comment]
This seems to be the last blocking issue, for 2.4.7 release. I can start the process once this is merged to branch-2.4.

**Comment by cloud-fan** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 8]
```diff
@@ -1237,20 +1250,42 @@ class Analyzer(
...
+        }
+
+        if (attrMapping.isEmpty) {
+          newPlan -> attrMapping
+        } else {
+          assert(!attrMapping.groupBy(_._1.exprId)
+            .exists(_._2.map(_._2.exprId).distinct.length > 1),
+            "Found duplicate rewrite attributes")
+          val attributeRewrites = AttributeMap(attrMapping)
+          // rewrite the attributes of parent node
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1282]: `newPlan` is exactly the current node. So `parent node` reads confusing.

How about "Using attrMapping from the children plans to rewrite their parent node. Note that we shouldn't rewrite a node using attrMapping from its sibling nodes."

**Review by viirya — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 8 above)
Comment (Ngone51) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1282]: Looks better, thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 7 above)
Comment (Ngone51) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1264]: Do you mean `conflictPlanMap(plan) -> plan.output.zip(conflictPlanMap(plan).output)`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 9]
```diff
@@ -1237,20 +1250,43 @@ class Analyzer(
...
+          attrMapping ++= childAttrMapping
+          newChild
+        }
+
+        if (attrMapping.isEmpty) {
+          newPlan -> attrMapping
+        } else {
+          assert(!attrMapping.groupBy(_._1.exprId)
+            .exists(_._2.map(_._2.exprId).distinct.length > 1),
+            "Found duplicate rewrite attributes")
```
Comment (imback82) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1280]: When the query from [SPARK-32280](https://issues.apache.org/jira/browse/SPARK-32280) is run, I believe this assert is hit. Should this be an assert or should we throw an AnalysisException?

Comment (Ngone51) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1280]: Thanks for the reference. Yeah, it hits the assert. I'm looking into it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 7 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1264]: ah i see, nvm.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 10]
```diff
@@ -1237,20 +1250,79 @@ class Analyzer(
...
+          //          :           +- Project [1 AS id#227]
+          //          :              +- OneRowRelation
+          //          +- SubqueryAlias r
+          //             +- SubqueryAlias b
+          //                +- Project [id#224, foo AS kind#223]
+          //                   +- SubqueryAlias a
+          //                      +- Project [1 AS id#224]
+          //                         +- OneRowRelation
+          attrMapping ++= childAttrMapping.filter { case (oldAttr, _) =>
+            (plan.references ++ plan.outputSet ++ plan.producedAttributes).contains(oldAttr)
```
Comment (Ngone51) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1306]: Hi all @cloud-fan @maropu @viirya , please take a look at the latest change according to SPARK-32280 meitoned by @imback82 . I've also added a new test migrated from SPARK-32280.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1306]: Can you add [SPARK-32280] in the PR title too?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 11]
```diff
@@ -1237,20 +1250,79 @@ class Analyzer(
...
+      } else {
+        val attrMapping = new mutable.ArrayBuffer[(Attribute, Attribute)]()
+        val newPlan = plan.mapChildren { child =>
+          // If not, we'd rewrite child plan recursively until we find the
+          // conflict node or reach the leaf node.
+          val (newChild, childAttrMapping) = rewritePlan(child, conflictPlanMap)
+          // Only return rewrite attributes which could be used by the parent node.
+          // Otherwise, it could introduce duplicate rewrite attributes. For example,
+          // for the following plan, if we don't do filter for the `childAttrMapping`,
+          // the node `SubqueryAlias b` will return rewrite attribute of [kind#220 -> kind#228]
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1274]: You didn't explain `kind#20` ahead. We should probably explain what's in the `conflictPlanMap` at the beginning of the comment

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1274]: And can we use attr id `1, 2, 3, 4, ...` instead?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 10 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1306]: why do we need `plan.producedAttributes`?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1306]: and seems we only need `plan.outputSet`: if an attribute is not output by the current plan, it's useless to propagate this attribute to the parent nodes.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1306]: The above is probably good enough as the comment here.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1306]: Seems we need `plan.outputSet` too. I think it is similar to case of aliases in projection above you added.

Comment (Ngone51) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1306]: `plan.references` is also necessary? For example,  Join's condition references the attributes but its output doesn't contain the attribute.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 11 above)
Comment (Ngone51) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: I've changed to the short comment as you mentioned above.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 10 above)
Comment (Ngone51) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: > Seems we need plan.outputSet too. I think it is similar to case of aliases in projection above you added.

Yeah, "aliases in projection" is another case. And I assume you were referring `plan.references` too. @viirya

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: `plan.references` won't be the output, right? I guess they won't conflict with others?

Comment (Ngone51) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Yes, they won't. I mean, you probably have a typo above:

 > Seems we need plan.outputSet too.

IIUC, it sounds like you're actually meaning the ` plan.references` according to the context.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 12]
```diff
@@ -1237,20 +1250,48 @@ class Analyzer(
...
+        val attrMapping = new mutable.ArrayBuffer[(Attribute, Attribute)]()
+        val newPlan = plan.mapChildren { child =>
+          // If not, we'd rewrite child plan recursively until we find the
+          // conflict node or reach the leaf node.
+          val (newChild, childAttrMapping) = rewritePlan(child, conflictPlanMap)
+          attrMapping ++= childAttrMapping.filter { case (oldAttr, _) =>
+            // If an attribute from the child is not output or referenced by the parent plan,
+            // it's useless to propagate this attribute to the parent plan, as they are not
+            // going to use this attribute.
+            (plan.outputSet ++ plan.references).contains(oldAttr)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1275]: the parents of `plan` can only reference `plan.outputSet`, that's why I think only `plan.outputSet` is necessary.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1275]: ah sorry, you are right. `attrMapping` is not only propagated to parents, but is also used to rewrite attributes of the currennt `plan`, so `plan.references` is necessary. Can we make the comment clearer about it?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1275]: maybe something like
```
// `attrMapping` is used to replace the attributes of the current `plan`, and also to be propagated to
// parent plans, so the `oldAttr` must either be part of `plan.references` (so that it can be used to
// replace attributes of current plan) or `plan.outputSet` (so that it can be used by parent plans).
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 10 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Nvm, I actually referred `plan.outputSet`. But now I see `plan.references` is also needed.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 13]
```diff
@@ -3565,13 +3565,36 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
     withTable("t") {
       sql("CREATE TABLE t USING PARQUET AS SELECT 1 AS id")
       checkAnswer(
-        sql(s"""
-          |WITH cte AS (SELECT /*+ REPARTITION(3) */ * FROM t)
-          |SELECT * FROM cte
+        sql(
+          s"""
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:3569]: can you revert the unrelated change?

Comment (Ngone51) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: oops..

**Comment by cloud-fan** [comment]
The last commit is just a code style change, I'm merging it to master/3.0, thanks!

**Comment by cloud-fan** [comment]
@Ngone51 can you send a backport PR for 2.4? thanks!

**Comment by Ngone51** [comment]
Sure! Thanks all!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 14]
```diff
@@ -1237,20 +1250,50 @@ class Analyzer(
...
+            // the `oldAttr` must be part of either `plan.references` (so that it can be used to
+            // replace attributes of the current `plan`) or `plan.outputSet` (so that it can be
+            // used by those parent plans).
+            (plan.outputSet ++ plan.references).contains(oldAttr)
+          }
+          newChild
+        }
+
+        if (attrMapping.isEmpty) {
+          newPlan -> attrMapping
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1283]: cc @srowen and @HyukjinKwon  . `catalyst` module compilation on `Scala 2.13` is broken here.
```
[ERROR] [Error] /Users/dongjoon/PRS/SPARK-testCommandAvailable/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1284: type mismatch;
 found   : scala.collection.mutable.ArrayBuffer[(org.apache.spark.sql.catalyst.expressions.Attribute, org.apache.spark.sql.catalyst.expressions.Attribute)]
 required: Seq[(org.apache.spark.sql.catalyst.expressions.Attribute, org.apache.spark.sql.catalyst.expressions.Attribute)]
```

Comment (srowen) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1283]: Yep, I'm sure we're going to have to go back and add fixes for new code that's added while we're working on 2.13, which has similar issues. This just requires a `.toSeq`. I think you can tack this on to other fixes you make in the module along the way -- just fix it when it blocks progress.

Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1283]: +1 for that approach. Thanks.

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:1283]: Thanks for letting me know. +1
